File: app/src/main/java/com/abdurazaaqmohammed/AntiSplit/main/MainActivity.java
package com.abdurazaaqmohammed.AntiSplit.main;

import static android.provider.Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION;
import static com.reandroid.apkeditor.merge.LogUtil.logEnabled;

import android.Manifest;
import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.provider.OpenableColumns;
import android.support.v4.content.FileProvider;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.IntentSenderRequest;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.WindowCompat;
import androidx.core.widget.NestedScrollView;

import com.abdurazaaqmohammed.AntiSplit.R;
import com.corentinc.patcher.CPatcher;
import com.fom.storage.media.AndroidXI;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.reandroid.apk.ApkBundle;
import com.reandroid.apkeditor.merge.LogUtil;
import com.reandroid.apkeditor.merge.Merger;
import com.starry.FileUtils;

import java.io.File;
import java.lang.ref.WeakReference;
import java.nio.channels.ClosedByInterruptException;
import java.util.Objects;

public class MainActivity extends AppCompatActivity implements Merger.LogListener {
    public static boolean errorOccurred;
    public static String lang;
    public static int theme;
    public DeviceSpecsUtil DeviceSpecsUtil;
    private String pkgName;

    public Handler getHandler() {
        return handler;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        handler = new Handler(Looper.getMainLooper());

        deleteDir(getCacheDir());
        WindowCompat.setDecorFitsSystemWindows(getWindow(), false);
        DeviceSpecsUtil = new DeviceSpecsUtil(this);
        setContentView(R.layout.activity_main);
        scrollView = findViewById(R.id.scrollView);
        logField = findViewById(R.id.logField);
        LogUtil.setLogListener(this);
        logEnabled = true;
        pkgName = "com.spotify.music";
        process(Uri.fromFile(new File(getApplicationContext().getCacheDir().getPath() + "spotify.apk")));
    }

    public void styleAlertDialog(AlertDialog ad) {
        Window w = ad.getWindow();
        if (w != null) {
            GradientDrawable border = new GradientDrawable();
            border.setColor(theme == com.google.android.material.R.style.Theme_Material3_Light_NoActionBar ? Color.WHITE : Color.BLACK); // Background color
            TypedValue typedValue = new TypedValue();
            getTheme().resolveAttribute(android.R.attr.colorPrimary, typedValue, true);
            border.setStroke(5, typedValue.data); // Border width and color
            border.setCornerRadius(24);
            w.setBackgroundDrawable(border);
            double m = 0.8;
            DisplayMetrics displayMetrics = this.getResources().getDisplayMetrics();
            int height = (int) (displayMetrics.heightPixels * m);
            int width = (int) (displayMetrics.widthPixels * m);
            w.setLayout(width, height);
        }
        runOnUiThread(ad::show);
    }

    private void checkStoragePerm() {
        if (doesNotHaveStoragePerm(this)) {
            Toast.makeText(this, this.getString(R.string.grant_storage), Toast.LENGTH_LONG).show();
            if (LegacyUtils.supportsWriteExternalStorage)
                requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 0);
            else
                startActivityForResult(new Intent(ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION, Uri.parse("package:" + getPackageName())), 0);
        }
    }

    public static boolean doesNotHaveStoragePerm(Context context) {
        return (LegacyUtils.supportsWriteExternalStorage ?
                context.checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_DENIED :
                !Environment.isExternalStorageManager());
    }


    TextView logField;
    NestedScrollView scrollView;

    @Override
    public void onLog(CharSequence msg) {
        onLog(msg.toString());
    }

    @Override
    public void onLog(String log) {
        Log.i("", log);
        runOnUiThread(() -> {
            logField.append(new StringBuilder(log).append('\n'));
            scrollView.post(() -> scrollView.fullScroll(View.FOCUS_DOWN));
        });
    }

    @Override
    public void onLog(int resID) {
        onLog(this.getString(resID));
    }

    private Handler handler;

    /**
     * @noinspection ResultOfMethodCallIgnored, DataFlowIssue
     */
    public static void deleteDir(File dir) {
        // There should never be folders in here.
        for (String child : dir.list()) new File(dir, child).delete();
    }

    @Override
    protected void onDestroy() {
        deleteDir(getCacheDir());
        super.onDestroy();
    }

    private static class ProcessTask extends AsyncTask<Uri, Void, Void> {
        private final WeakReference<MainActivity> activityReference;
        private final String packageNameFromAppList;

        // only retain a weak reference to the activity
        ProcessTask(MainActivity context, String fromAppList) {
            activityReference = new WeakReference<>(context);
            this.packageNameFromAppList = fromAppList;
        }

        @Override
        protected Void doInBackground(Uri... uris) {
            MainActivity activity = activityReference.get();
            if (activity == null) return null;

            final File cacheDir = activity.getCacheDir();
            deleteDir(cacheDir);
            try {
                ApkBundle bundle = new ApkBundle();
                bundle.loadApkDirectory(new File(activity.getPackageManager().getPackageInfo(packageNameFromAppList, 0).applicationInfo.sourceDir).getParentFile(), false, activity);
                Merger.run(bundle, cacheDir, uris[0], activity, true);
            } catch (Exception e) {
                activity.showError(e);
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void result) {
            MainActivity activity = activityReference.get();
            activity.pkgName = null;
            activity.showSuccess();
        }
    }

    public static String getOriginalFileName(Context context, Uri uri) {
        String result = null;
        try {
            if (Objects.equals(uri.getScheme(), "content")) {
                try (Cursor cursor = context.getContentResolver().query(uri, null, null, null, null)) {
                    if (cursor != null && cursor.moveToFirst()) {
                        result = cursor.getString(cursor.getColumnIndexOrThrow(OpenableColumns.DISPLAY_NAME));
                    }
                }
            }
            if (result == null) {
                result = uri.getPath();
                int cut = Objects.requireNonNull(result).lastIndexOf('/'); // Ensure it throw the NullPointerException here to be caught
                if (cut != -1) result = result.substring(cut + 1);
            }
            LogUtil.logMessage(result);
            String suffix = "_antisplit";
            return result.replaceFirst("\\.(?:xapk|aspk|apk[sm])", suffix + ".apk");
        } catch (Exception ignored) {
            return "filename_not_found";
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && data != null && requestCode == 0) {
            checkStoragePerm();
        }
    }

    private final ActivityResultLauncher<IntentSenderRequest> launcher = registerForActivityResult(
            new ActivityResultContracts.StartIntentSenderForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) LogUtil.logMessage("Deleted ");
            });

    private void process(Uri outputUri) {
        findViewById(R.id.installButton).setVisibility(View.GONE);
        ProcessTask processTask = new ProcessTask(this, pkgName);
        processTask.execute(outputUri);
        LinearLayout fabs = findViewById(R.id.fabs);
        fabs.setAlpha(0.5f);
        View cancelButton = findViewById(R.id.cancelButton);
        cancelButton.setVisibility(View.VISIBLE);
        cancelButton.setOnClickListener(v -> {
            try {
                if (doesNotHaveStoragePerm(this))
                    AndroidXI.getInstance().with(this).delete(launcher, outputUri);
                else if (new File(FileUtils.getPath(outputUri, this)).delete())
                    LogUtil.logMessage("Cleaned output file " + getOriginalFileName(this, outputUri));
            } catch (Exception ignored) {
            }
            Intent intent = getPackageManager().getLaunchIntentForPackage(getPackageName());
            if (intent == null) {
                processTask.cancel(true);
                intent = getIntent();
                finish();
                startActivity(intent);
            } else {
                startActivity(Intent.makeRestartActivityTask(intent.getComponent()));
                Runtime.getRuntime().exit(0);
            }
        });

        View copyButton = findViewById(R.id.copyButton);
        copyButton.setVisibility(View.VISIBLE);
        copyButton.setOnClickListener(v -> copyText(new StringBuilder().append(logField.getText()).append('\n').append(((TextView) findViewById(R.id.errorField)).getText())));
    }

    private void showSuccess() {
        //patch here
        findViewById(R.id.cancelButton).setVisibility(View.GONE);

        View installButton = findViewById(R.id.installButton);
        if (errorOccurred) installButton.setVisibility(View.GONE);
        else {
            final String success = this.getString(R.string.success_saved);
            LogUtil.logMessage(success);
            runOnUiThread(() -> Toast.makeText(this, success, Toast.LENGTH_SHORT).show());
            if (Merger.signedApk != null) {
                File patchedApk = CPatcher.INSTANCE.patch(getApplicationContext(), Merger.signedApk, this);
                installButton.setVisibility(View.VISIBLE);
                installButton.setOnClickListener(v ->
                        startActivity(new Intent(Intent.ACTION_INSTALL_PACKAGE)
                                .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                                .setData(FileProvider.getUriForFile(getApplicationContext(), "com.abdurazaaqmohammed.AntiSplit.provider", patchedApk))));
            } else installButton.setVisibility(View.GONE);
        }
    }

    private void copyText(CharSequence text) {
        ((ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE)).setPrimaryClip(ClipData.newPlainText("log", text));
        Toast.makeText(this, this.getString(R.string.copied_log), Toast.LENGTH_SHORT).show();
    }

    private void showError(Exception e) {
        if (!(e instanceof ClosedByInterruptException)) {
            final String mainErr = e.toString();
            errorOccurred = !mainErr.equals(this.getString(R.string.sign_failed));

            StringBuilder stackTrace = new StringBuilder(mainErr);

            for (StackTraceElement line : e.getStackTrace()) stackTrace.append(line).append('\n');
            StringBuilder fullLog = new StringBuilder(stackTrace).append('\n')
                    .append("SDK ").append(Build.VERSION.SDK_INT).append('\n')
                    .append(this.getString(R.string.app_name)).append(' ');
            String currentVer;
            try {
                currentVer = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
            } catch (Exception ex) {
                currentVer = "2.1.1";
            }
            fullLog.append(currentVer).append('\n').append("Storage permission granted: ").append(!doesNotHaveStoragePerm(this))
                    .append('\n').append(logField.getText());

            getHandler().post(() -> runOnUiThread(() -> {
                View dialogView = getLayoutInflater().inflate(R.layout.dialog_button_layout, null);

                ((TextView) dialogView.findViewById(R.id.errorD)).setText(stackTrace);

                styleAlertDialog(new MaterialAlertDialogBuilder(this)
                        .setTitle(mainErr)
                        .setView(dialogView)
                        .setPositiveButton(this.getString(R.string.copy_log), (dialog, which) -> {
                            copyText(fullLog);
                            dialog.dismiss();
                        })
                        .setNegativeButton(this.getString(R.string.create_issue), (dialog, which) -> {
                            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/AbdurazaaqMohammed/AntiSplit-M/issues/new?title=Crash%20Report&body=" + fullLog)));
                            dialog.dismiss();
                        })
                        .setNeutralButton(this.getString(R.string.cancel), (dialog, which) -> dialog.dismiss())
                        .create());
                ScrollView scrollView = dialogView.findViewById(R.id.errorView);

                ViewGroup.LayoutParams params = scrollView.getLayoutParams();
                params.height = (int) (this.getResources().getDisplayMetrics().heightPixels * 0.5);
                scrollView.setLayoutParams(params);
            }));
        }
    }
}

File: app/src/main/java/com/abdurazaaqmohammed/AntiSplit/main/MainActivity.kt
package com.abdurazaaqmohammed.AntiSplit.main

import android.Manifest
import android.app.ActivityManager
import android.content.ClipData
import android.content.ClipboardManager
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.PackageManager.NameNotFoundException
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.util.TypedValue
import android.view.View
import android.view.WindowManager
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import androidx.core.view.WindowCompat
import androidx.lifecycle.lifecycleScope
import com.corentinc.patcher.AppInstaller
import com.corentinc.patcher.AppUpdater
import com.corentinc.patcher.clearDirectory
import com.corentinc.patcher.isNetworkException
import com.corentinc.patcher.saveToDownloadsFolder
import com.corentinc.screens.patcher.ui.AutoPatcherScreen
import com.github.corentinc.SpotifyAutoPatcher.R
import com.github.corentinc.httpcodescats.ui.theme.AutoPatcherTheme
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.reandroid.apkeditor.merge.LogUtil
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File
import java.nio.channels.ClosedByInterruptException
import java.util.Calendar
import java.util.zip.ZipException
import kotlin.io.path.createDirectory

const val PACKAGE_TO_PATCH = "com.google.android.apps.youtube.music"
private const val TEMP_FOLDER = "temp"

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {
	private lateinit var defaultFolder: File

	private val requestWritePermissionLauncher = registerForActivityResult(
		ActivityResultContracts.RequestPermission()
	) { isGranted: Boolean ->
		if (!isGranted) {
			runOnUiThread {
				showAlertDialog(
					getString(R.string.storage_permission_denied_warning),
					positiveButtonText = getString(R.string.ok),
					positiveButtonAction = {
						// empty
					},
				)
			}
		}
	}

	private val uninstallCallback =
		registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { _ ->
			patchedApk?.let {
				installAppOrShowPopUpIfAlreadyInstalled(it)
			}
		}


	private var logField: TextView? = null

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		enableEdgeToEdge()

		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q && ContextCompat.checkSelfPermission(
				this,
				Manifest.permission.WRITE_EXTERNAL_STORAGE
			) != PackageManager.PERMISSION_GRANTED
		) {
			requestWritePermissionLauncher.launch(Manifest.permission.WRITE_EXTERNAL_STORAGE)
		}
		setContent {
			AutoPatcherTheme {
				AutoPatcherScreen(
					title = getString(R.string.app_name),
					onPatchingFinished = {
						showAlertDialog(
							getString(R.string.ready_to_install),
							positiveButtonText = getString(R.string.next),
							positiveButtonAction = {
								AppInstaller.uninstallApp(uninstallCallback)
							},
						)
					},
					onCopyClick = { text ->
						copyText(
							text
						)
					},
					onInstallClick = { patch ->
						patch?.let {
							patchedApk = patch
							installAppOrShowPopUpIfAlreadyInstalled(patch)
						}
					},
					onCancelClick = {
						restartActivity()
					},
					onDownloadClick = { patch ->
						patch?.let {
							patchedApk = patch
							installAppOrShowPopUpIfAlreadyInstalled(patch)
						}
					},
					onError = { error ->
						showError(error)
					},
					defaultFolder = defaultFolder
				)
			}
		}

		defaultFolder = File(cacheDir, TEMP_FOLDER)
		if (!defaultFolder.exists()) defaultFolder.toPath().createDirectory()
		defaultFolder.clearDirectory()
		WindowCompat.setDecorFitsSystemWindows(window, false)

		LogUtil.logEnabled = true

		lifecycleScope.launch(Dispatchers.IO) {
			AppUpdater.checkIfAnUpdateIsAvailable(
				this@MainActivity,
				defaultFolder,
				onUpdateAvailable = { latestVersionApk ->
					showAlertDialog(
						getString(R.string.a_new_version_of_spotifyautopatcher_is_available_do_you_want_to_install_it),
						positiveButtonText = getString(R.string.install),
						positiveButtonAction = {
							AppInstaller.installApp(this@MainActivity, latestVersionApk)
							showStartProcessDialog()

						},
						neutralButtonText = getString(R.string.skip),
						neutralButtonAction = {
							showStartProcessDialog()
						}
					)
				},
				onUpdateNotAvailable = {
					LogUtil.logMessage(getString(R.string.no_update_available))
					showStartProcessDialog()
				}).onFailure { exception ->
				LogUtil.logMessage(
					getString(
						R.string.could_not_check_for_spotifyautopatcher_updates,
						exception
					)
				)
				showStartProcessDialog()
			}
		}
	}

	private fun showStartProcessDialog() {
		showAlertDialog(
			getString(R.string.before_start_message),
			positiveButtonText = getString(R.string.start),
			positiveButtonAction = {
				//mergeAndPatchApk()
			},
		)
	}

	private fun showAlertDialog(
		text: String,
		positiveButtonText: String,
		positiveButtonAction: () -> Unit,
		neutralButtonText: String? = null,
		neutralButtonAction: (() -> Unit)? = null,
	) {
		runOnUiThread {
			val builder = MaterialAlertDialogBuilder(this)
			builder.setCancelable(false)
			builder.setMessage(text)
			builder.setPositiveButton(
				positiveButtonText
			) { dialog: DialogInterface, _: Int ->
				positiveButtonAction()
				dialog.dismiss()
			}
			neutralButtonText?.let {
				neutralButtonAction?.let {
					builder.setNeutralButton(
						neutralButtonText
					) { dialog: DialogInterface, _: Int ->
						neutralButtonAction()
						dialog.dismiss()
					}
				}
			}
			styleAlertDialog(builder.create())
		}
	}

	private fun styleAlertDialog(ad: AlertDialog) {
		val w = ad.window
		if (w != null) {
			val border = GradientDrawable()
			val typedValue = TypedValue()
			theme.resolveAttribute(android.R.attr.colorPrimary, typedValue, true)
			border.setStroke(5, typedValue.data) // Border width and color
			border.cornerRadius = 24f
			w.setBackgroundDrawable(border)
			val m = 0.8
			val displayMetrics = this.resources.displayMetrics
			val height = (displayMetrics.heightPixels * m).toInt()
			val width = (displayMetrics.widthPixels * m).toInt()
			w.setLayout(width, height)
		}
		runOnUiThread { ad.show() }
	}

	override fun onResume() {
		super.onResume()
		val window = this.window
		window?.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
	}

	override fun onPause() {
		super.onPause()
		val window = this.window
		window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
	}


	private fun saveApk(apk: File) {
		apk.saveToDownloadsFolder(
			contentResolver,
			"Spotify(SpotifyAutoPatcher)-" + Calendar.getInstance().timeInMillis + ".apk"
		)
			.onFailure {
				showAlertDialog(
					getString(R.string.could_not_save_apk),
					positiveButtonText = getString(R.string.ok),
					positiveButtonAction = {
						// empty
					},
				)
			}
			.onSuccess {
				val downloadsDirectory = Environment.getExternalStoragePublicDirectory(
					Environment.DIRECTORY_DOWNLOADS
				)
				showAlertDialog(
					getString(R.string.apk_saved) + " : ${downloadsDirectory.path}",
					positiveButtonText = getString(R.string.ok),
					positiveButtonAction = {
						// empty
					},
				)
			}
	}

	private fun restartActivity() {
		var intent = packageManager.getLaunchIntentForPackage(packageName)
		if (intent == null) {
			intent = getIntent()
			finish()
			startActivity(intent)
		} else {
			startActivity(Intent.makeRestartActivityTask(intent.component))
			Runtime.getRuntime().exit(0)
		}
	}

	private fun installAppOrShowPopUpIfAlreadyInstalled(patchedApk: File) {
		try {
			this@MainActivity.packageManager.getPackageInfo(
				PACKAGE_TO_PATCH, 0
			).applicationInfo!!.sourceDir
			showAlertDialog(
				getString(R.string.spotify_detected_before_install),
				positiveButtonText = getString(R.string.ok),
				positiveButtonAction = {
					// empty
				},
				neutralButtonText = getString(R.string.install_anyway_not_recommended),
				neutralButtonAction = {
					AppInstaller.installApp(this, patchedApk)
				}
			)
		} catch (exception: NameNotFoundException) {
			AppInstaller.installApp(this, patchedApk)
		}
	}

	private var patchedApk: File? = null

	private fun copyText(text: CharSequence) {
		(getSystemService(CLIPBOARD_SERVICE) as ClipboardManager).setPrimaryClip(
			ClipData.newPlainText(
				"log",
				text
			)
		)
		Toast.makeText(
			this,
			this.getString(R.string.copied_log), Toast.LENGTH_SHORT
		).show()
	}

	private fun showError(error: Throwable) {
		when {
			error.isNetworkException() -> {
				showAlertDialog(
					getString(R.string.network_unavailable_error),
					positiveButtonText = getString(R.string.retry),
					positiveButtonAction = {
						restartActivity()
					},
				)
			}

			error is NameNotFoundException -> {
				showAlertDialog(
					getString(R.string.app_not_found_error),
					positiveButtonText = getString(R.string.retry),
					positiveButtonAction = {
						restartActivity()
					},
				)
			}

			error is ZipException -> {
				showAlertDialog(
					getString(R.string.zip_exception_error),
					positiveButtonText = getString(R.string.fix),
					positiveButtonAction = {
						(getSystemService(ACTIVITY_SERVICE) as ActivityManager).clearApplicationUserData()
					},
					neutralButtonText = getString(R.string.retry),
					neutralButtonAction = {
						restartActivity()
					}
				)
			}

			error !is ClosedByInterruptException -> {
				val mainErr = error.toString()

				val stackTrace = StringBuilder(mainErr)

				for (line in error.stackTrace) stackTrace.append(line).append('\n')
				val fullLog = StringBuilder(stackTrace).append('\n')
					.append("SDK ").append(Build.VERSION.SDK_INT).append('\n')
					.append(this.getString(R.string.app_name)).append(' ')
				val currentVer = packageManager.getPackageInfo(packageName, 0).versionName
				fullLog.append(currentVer).append('\n').append("Storage permission granted: ")
					.append('\n').append(logField!!.text)

				runOnUiThread {
					val dialogView = layoutInflater.inflate(
						R.layout.dialog_button_layout,
						null
					)
					(dialogView.findViewById<View>(R.id.errorD) as TextView).text =
						stackTrace

					styleAlertDialog(
						MaterialAlertDialogBuilder(this)
							.setTitle(mainErr)
							.setCancelable(false)
							.setView(dialogView)
							.setPositiveButton(
								this.getString(R.string.copy_log)
							) { dialog: DialogInterface, _: Int ->
								copyText(fullLog)
								dialog.dismiss()
							}
							.setNegativeButton(
								this.getString(R.string.create_issue)
							) { dialog: DialogInterface, _: Int ->
								startActivity(
									Intent(
										Intent.ACTION_VIEW,
										"https://github.com/corentin-c/SpotifyAutoPatcher/issues/new?title=Crash%20Report&body=$fullLog".toUri()
									)
								)
								dialog.dismiss()
							}
							.setNeutralButton(
								this.getString(R.string.cancel)
							) { dialog: DialogInterface, _: Int -> dialog.dismiss() }
							.create())
					val scrollView =
						dialogView.findViewById<ScrollView>(R.id.errorView)

					val params = scrollView.layoutParams
					params.height =
						(this.resources.displayMetrics.heightPixels * 0.5).toInt()
					scrollView.layoutParams = params
				}
			}
		}
	}
}